### ISSUE TITLE
How to use EGNN to update edge properties based on node features and coordinates?

### ISSUE THREAD

**Issue by Wes-V3** [lucidrains/egnn-pytorch#40]
Hello, I am using your EGNN implementation for a project involving protein structure analysis, where I am representing residues as nodes. Currently, I have each residue type as a feature, and each residue also has corresponding coordinates.

What I am trying to achieve is to update all the edge properties based on the residue type and the coordinates. My goal is to adjust the interactions between the residues based on these properties, but I am not sure about the best way to go about this.

I've looked through the current documentation and examples but couldn't find an example or guideline about how to update edge properties in this way. I am not sure how to proceed and would appreciate any guidance you can provide.

**Comment by lucidrains** [comment]
oh hey Sen

that's a bit outside the scope of the original paper

but i think it is doable outside the module, if you simply project the output node representations into `n_i` and `n_o`, do an outer sum `n_i[..., None, :] + n_o[..., :, None]` and then a small MLP to your final prediction? (perhaps even concat the relative distance pre-MLP)?

**Comment by lucidrains** [comment]
@Wes-V3 another way would be to save all your attention maps and project along the merged head and depth axis

**Comment by Wes-V3** [comment]
> oh hey Sen
> 
> that's a bit outside the scope of the original paper
> 
> but i think it is doable outside the module, if you simply project the output node representations into `n_i` and `n_o`, do an outer sum `n_i[..., None, :] + n_o[..., :, None]` and then a small MLP to your final prediction? (perhaps even concat the relative distance pre-MLP)?

Thank you for your guidance. Your suggestion seems promising. I've given it a shot, and also experimented with the method recommended by PyTorch_Geometric [How to update edges?](https://github.com/pyg-team/pytorch_geometric/issues/813):

```
row, col = edge_index
new_edge_attr = self.mlp(torch.cat([x[row], x[col], edge_attr], dim=-1))
```
This approach appears to align better with my particular use case.

Despite these efforts, the performance has not reached the desired level. I am continuously exploring ways to improve and refine the solution. Any additional insights you may have would be greatly appreciated.

Thank you once again for your support.
